[C#] Extract concepts from v2 change exercise (#511)
exercism/v3: languages/csharp/reference/exercise-concepts/change.md

# Concepts of change

[Example implementation](https://github.com/exercism/csharp/blob/master/exercises/change/Example.cs)

## General

- functions: used as the main entry point for the exercise

> - function arguments: input strand is passed as an arguments

Reviewer: ```suggestion
- function arguments: input coins and target are passed as arguments
```

- out parameters: used when retrieving a value from the dictionary
- return values: returning a value from a method
- conditionals using if: conditionally execute logic using an `if` statement and the ternary operator
- exceptions: throw an exception in the event of invalid input
- scoping: use `{` and `}` to denote scoping
- classes: the tested method is defined in a class
- objects: creating an object to keep track of the coins
- members: methods linked to a class instance
- visibility: making tested method and tested class `public`
- imports: import types through `using` statements
- namespaces: knowing where to find the `List<T>` and `Dictionary<TKey, TValue>` classes

> - type inference: using `var` to define the score

Reviewer: ```suggestion
- type inference: using `var` to define the complex types (dictionaries)
```

> - assignment: assigning values, such as the score

Reviewer: ```suggestion
- assignment: assigning values, such as the minimalCoinsMap
```

> - mutation: mutating the list after it has been created

Reviewer: ```suggestion
- mutation: mutating the dictionary after it has been created
```

- mapping, selecting, ordering enumerables: using multiple LINQ methods to manipulate enumerables (`Min`, `Where`, `OrderBy`, `Select`, `Aggregate`, `FirstOrDefault`, `ToList`, `ToArray`)
- ranges: a range is created for iterating over all values up to the target
- object initializers: initializing dictionary through object initializer
- integers: an `int` is used for a coin and the target
- array: a `int []` is used for the available coinage

> - lists: a `List<int>` is used as to store the rolls in

Reviewer: ```suggestion
- lists: a `List<int>` is used as to store the coins for a given value in
```
